### PR TITLE
Make find_func use rom_table_lookup_fn

### PR DESCRIPTION
**Pull request by AsafFisher**

### CODE REVIEW

File: src/bootrom.rs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+// Create a ROMFuncs struct using the lookup function and the bootrom function's code.
+macro_rules! rom_funcs {
+    ($arg:expr, $lookup_func:expr, $($name:ident, $code:expr;) +) => {
+     ROMFuncs{
+         $($name: mem::transmute($lookup_func($arg, mem::transmute($code)))), +
+     }
+ }
+}
+
+macro_rules! ptr_def {
```
Comment (Dirbaio) [inline src/bootrom.rs]: You can omit the type in casts, so you can do this:

```rust
const ROM_TABLE_LOOKUP_PTR: *const u16 = 0x18 as _;
```

I think this is much more readable than introducing custom syntax with a macro

File: src/bootrom.rs [hunk 2]
```diff
@@ -0,0 +1,79 @@
+use core::mem;
+
+// A bootrom function table code.
+pub type RomFnTableCode = [u8; 4];
```
Comment (Dirbaio) [inline src/bootrom.rs]: Table codes are 2 chars, not 4.

**Review by Dirbaio — CHANGES_REQUESTED** [review]
I agree with changing this to use the lookup fn from ROM, I didn't notice that was an option. 

However I feel introducing macros hurts readability a lot. Why not simply changing `find_func` to call the ROM lookup func?

**Comment by AsafFisher** [comment]
My thoughts were that macros make code more extensible and while it’s true that the macro itself is not very friendly adding definitions become much easier 

At the beginning I actually wanted to make a macro called rom_function! 
That works like that:


```rust
rom_function!(map_page, page: ptr, page_mode: *const u16,..., MP)

// It follows 
rom_function!(symbol, arg..., CODE)
```

But rust is too young for that... are you sure it’s so bad to read? I think it actually makes it more readable as an abstraction but indeed that macro implementation is not friendly

File: src/bootrom.rs (hunk 2 above)
Comment (AsafFisher) [reply src/bootrom.rs]: The code themselves are 2 chars but if you take a look at macro `rom_table_code` 'bootrom.h' in the SDK you will see that it's actually 32 bit (: 
The 8 first bits (from the right) are the second letter and the other 8 bits are the first letter. 
The rest 16 bits are zero (maybe not need to read the schema)

File: src/bootrom.rs (hunk 1 above)
Comment (AsafFisher) [reply src/bootrom.rs]: Do you know why the `as _` is needed?

Comment (AsafFisher) [reply src/bootrom.rs]: Done and removed `ptr_def!` macro.

File: src/bootrom.rs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+const ROM_TABLE_LOOKUP_PTR: *const u16 = 0x18 as _;
+const OPS_TABLE_PTR: *const u16 = 0x14 as _;
+
+/// This function searches for (table)
+type RomTableLookupFn<T> = unsafe extern "C" fn(*const u16, u32) -> T;
+
+/// Given a rom_address pointer, convert a 16 bit pointer stored
+/// at the given rom address into a 32 bit pointer
+unsafe fn get_ptr_from_rom(rom_address: *const u16) -> *const u16 {
```
Comment (timokroeger) [inline src/bootrom.rs:15]: This function seems to be redundant

Comment (AsafFisher) [reply src/bootrom.rs:15]: It's not, it is explicit. This function should be inlined.

**Comment by AsafFisher** [comment]
Beautiful:
Screen Shot 2021-01-31 at 8 00 32 PM

File: src/bootrom.rs (hunk 3 above)
Comment (timokroeger) [reply src/bootrom.rs:15]: :+1: get it, missed the dereference

**Comment by AsafFisher** [comment]
I think I am going to close it since its not going to get in anytime soon and I already implemented it in the HAL project.

**Pull request closed without merging**